plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,340 @@
+import io.trino.spi.eventlistener.QueryContext;
+import io.trino.spi.eventlistener.QueryCreatedEvent;
+import io.trino.spi.eventlistener.QueryFailureInfo;
+import io.trino.spi.eventlistener.QueryIOMetadata;
+import io.trino.spi.eventlistener.QueryMetadata;
+import io.trino.spi.eventlistener.QueryOutputMetadata;
+import io.trino.spi.eventlistener.QueryStatistics;
+import io.trino.spi.resourcegroups.QueryType;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+
+public class OpenLineageListener
+        implements EventListener
+{
+    private final OpenLineage openLineage = new OpenLineage(URI.create("https://github.com/trinodb/trino/plugin/trino-openlineage"));
+    private final OpenLineageClient client;
+    private final String namespace;
+    private final Set<QueryType> includeQueryTypes;
+    private final ObjectMapper queryStatisticsMapper;
+    private static final Logger logger = Logger.get(OpenLineageListener.class);
+
+    @Inject
+    public OpenLineageListener(OpenLineageClient client, OpenLineageListenerConfig listenerConfig)
+            throws Exception
+    {
+        this.client = client;
+        this.namespace = listenerConfig.getTrinoURI().toString().replace(listenerConfig.getTrinoURI().getScheme(), "trino");
+
+        this.includeQueryTypes = listenerConfig.getIncludeQueryTypes();
+
+        this.queryStatisticsMapper = createQueryStatisticsObjectMapper();
+    }
+
+    @Override
+    public void queryCreated(QueryCreatedEvent queryCreatedEvent)
+    {
+        if (queryTypeSupported(queryCreatedEvent.getContext())) {
+            UUID runID = getQueryId(queryCreatedEvent.getMetadata());
+
+            try {
+                RunEvent event = getStartEvent(runID, queryCreatedEvent);
+                client.emit(event);
+            }
+            catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        else {
+            logger.debug(format("Query type %s not supported. Supported query types %s",
+                    queryCreatedEvent.getContext().getQueryType().toString(),
+                    this.includeQueryTypes));
+        }
+    }
+
+    @Override
+    public void queryCompleted(QueryCompletedEvent queryCompletedEvent)
+    {
+        if (queryTypeSupported(queryCompletedEvent.getContext())) {
+            UUID runID = getQueryId(queryCompletedEvent.getMetadata());
+
+            try {
+                RunEvent event = getCompletedEvent(runID, queryCompletedEvent);
+                client.emit(event);
+            }
+            catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
```

Comment:
Do we need this try-block, we don't see `getCompletedEvent` throws any `IllegalAccessException`